### ISSUE TITLE
Allow dispatch of bubbling custom events when using `@angular/elements`

### ISSUE THREAD

**Issue by yjaaidi** [angular/angular#39489]
# 🚀 feature request

### Relevant Package

This feature request is for `@angular/elements`.

### Description

When using `@angular/elements`, outputs are converted to `CustomEvents`:
https://github.com/angular/angular/blob/eb9eebbb45352da6000560ee0988d9d59dce1254/packages/elements/src/create-custom-element.ts#L225-L228

There is no way of listening to the custom events from a grand parent element or component because these custom events don't bubble as the `bubbles` option is set to `false` here:

https://github.com/angular/angular/blob/8fd25d96142ba5bd6674ba19848b5477e9bbe861/aio/src/testing/dom-utils.ts#L4-L16

It would be nice to figure out a way of listening to custom events in grand parent elements.

### Use Case
The app loads dynamic markdown content. This markdown is processed with a third party library like [Marked](https://github.com/markedjs/marked) where we add a special link handler that replaces links with our custom elements (using Angular Elements). These elements trigger custom events that we want to handle in the parent component (the one that loaded the markdown).

The general idea is that we can let a 3rd party (non-Angular) library control the DOM and inject our Angular Elements but we want to keep communicating with these elements without having to dispatch bubbling events manually or using an indirection like a provider or state management etc...

#### Solution A

1. Create a new `CustomEventEmitter` where we can configure `bubbles` and `cancelable`: 
```ts
new CustomEventEmitter({bubbles: true, cancelable: false})
```
_We can't use `EventEmitter` because `bubbles` and `cancelable` will be simply ignored and confusing when not using Angular elements_

2. Change `ComponentNgElementStrategy` in order to handle `CustomEventEmitter` differently instead of merging all emitters:
https://github.com/angular/angular/blob/eb9eebbb45352da6000560ee0988d9d59dce1254/packages/elements/src/component-factory-strategy.ts#L50

3. Change `createCustomElement` in order to handle `bubbles` and `cancelable`:
https://github.com/angular/angular/blob/eb9eebbb45352da6000560ee0988d9d59dce1254/packages/elements/src/create-custom-element.ts#L225-L228

4. Change `createCustomEvent` to allow the control of `bubbles` and `cancelable`:
https://github.com/angular/angular/blob/8fd25d96142ba5bd6674ba19848b5477e9bbe861/aio/src/testing/dom-utils.ts#L4-L16

5. [OPTIONAL] Should we deprecate `EventEmitter` usage with Angular elements? Allowing both can be confusing.

#### Solution B

Create a new `CustomEventEmitter` which is an `EventEmitter` of type `EventEmitter<{bubbles?: boolean, cancelable?: boolean, detail?: T}>` so every event can have a different bubbling strategy... but that would be a surprising behavior so I am not a big fan of this one.

#### Contribution
I'll be more than happy to contribute with a PR if the solution is suitable.

### Workaround
Trigger the event manually:
```ts
elementRef.nativeElement.dispatchEvent('myCustomEvent', {bubbles: true, detail: 'data'})
```

### Somewhat related issues
https://github.com/angular/angular/issues/2296

**Comment by gkalpak** [comment]
This sounds like a reasonable feature. If we want people to be able to turn Angular components to Custom Elements, we need to provide ways to implement common Custom Element behaviors (and supporting bubbling events is one of them imo).

##
As far as the implementation goes, I would prefer having an "additive" decorator (that would only take effect when `@angular/elements` is used) that would augment an `EventEmitter` with something like a `customEventConfig` property containing config for the `CustomEvent` (such as bubbles, cancelable, etc.).

Then, as @yjaaidi said, we should "teach" the `NgElementStrategy` to apply that config and use it when creating a `CustomEvent` for an `Output` emission.

The code could look something like this:

```diff
 // ...Other imports...
+import {CustomEventConfig} from '@angular/elements';
 
 @Component({ ... })
 export class MyComponent {
   @Output()
+  @CustomEventConfig({bubbles: true})
   myBubblyEvent = new EventEmitter<void>();
 
   // ...Other code...
 }
```

Everything else would remain the same from the app developer's perspective.

I believe this would be the least confusing option, especially when the component is intended to be used both as a regular component and as a Custom Element.

BTW, you can see this in action in [this very hacky POC][1]. (In lieu of the `CustomEventConfig` decorator, I am just adding a `customEventConfig` property to the `EventEmitter`.)

[1]: https://stackblitz.com/edit/ngelements-issue-39489?file=src%2Fapp%2Fapp.component.ts

**Comment by yjaaidi** [comment]
Thanks @gkalpak! You are right! This should be a decorator.
It just made me realize that we shouldn't assume that the output is an `EventEmitter` as it could be any observable.

```ts
class MyComponent {
  @Output()
  @CustomEventConfig({bubble: true})
  myBubblyEvent: Observable<void>();

  constructor() {
    myBubblyEvent = data$.pipe(...);
  }
}
```

I'll be happy to submit a PR if we are OK with the `CustomEventConfig` decorator.

**Comment by gkalpak** [comment]
Thx, @yjaaidi!

TBH, I am not aware of all the implications of adding a new decorator and other concerns there might arise.
So, let me check with the team and get back to you next week.

**Comment by gkalpak** [comment]
Sorry for the late response, @yjaaidi 😇

We discussed this and it seems a reasonable addition (which we would eventually want to get to). However, since this is something that affects the public API, there are several things we would need to consider/research (what is the right API to expose, how to future-proof it, what would the DX look like, how bundle sizes are affected, etc.). Unfortunately, there are higher priority efforts we are dedicating our resources towards atm; therefore this has to remain in the backlog for the time being.

Hopefully it will not be long before we can pick this up 🤞
<sub>
In the meantime, if you want to get involved in contributing, you are more than welcome to pick something from our [community-help][1] issue list :wink:
</sub>

[1]: https://github.com/angular/angular/issues?q=is%3Aissue+is%3Aopen+label%3A%22hotlist%3A+community-help%22

**Comment by yjaaidi** [comment]
No rush @gkalpak!

I'm waiting for the github notification! 😊

**Comment by PodgeHeavin** [comment]
This is a good feature request. We have started using `@angular/elements` as web components as a way to migrate from Aurelia to Angular. For bindings to work best in Aurelia bubbling of events is desirable from a performance point of view so to be able to optionally bubble events would be great.

**Comment by csakis** [comment]
This issue was raised over 6 years ago. It seems that everybody agrees that this is a desirable feature to have and @yjaaidi's solution is a good one. How long does it take to approve it  by the Angular team?

**Comment by PodgeHeavin** [comment]
This would be a very useful feature especially since we are mixing frameworks. Like Aurelia and Angular

**Comment by cgrabmann** [comment]
Hi,
I just stumbled over a similar issue as well.
I want to enable my consumers to call `preventDefault()` on the custom event, to stop my WebComponent from triggering an action. eg. navigation or submitting a form.

As a workaround I can of course provide my own preventDefault() method in the emitted value, which is then put on the details of the CustomEvent.
But I would find it more ergonomic that the consumers can use the standard Event-Api for this.

I agree that it would be nice to have the behavior if an output will bubble or not defined once.
But I think a nice solution would also be, to directly dispatch the value of an output if it is of type Event (or CustomEvent).

This would enable for all use-cases where you want to modify the CustomEvent dispatched by your WebComponent. If you don't use the component as WebComponent, but as normal Angular Component you can still just handle the emitted Event and have access to the same functionality.

Of course this has to big downsides:
1. this behavior is not obvious by looking at the public Angular API
2. the burden on making sure such attributes like `bubbles` are consistent between Event dispatches is on the implementing developer.
3. using the component as normal Angular Component would not have the exact same event behavior, since bubbling would not work.

Both downsides are not that big of a deal imo.
1. The new behavior can be documented next to the current documentation of wrapping the output value in a CustomEvent. Therefore the 2 options of how a output can be dispatched is found right next to the usage information, which should be sufficient.
2. This is a bit more tricky. But there could at least be a runtime error in dev-mode to hint that 2 consecutive events don't have the same attributes set.
3. This might be the biggest problem with this approach. Since that would lead to unintuitive behavior in these cases as you are handling a Event, which normally would behave differently from outputs.

Still doing it like this for now would already improve the situation and might not require as much design consideration as providing an explicit API for this use-case.   
A explicit API can be provided at a later point as well, if there is a need for a better DX.
